tempesta_fw/ss_skb.c

```diff
@@ -345,96 +369,106 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
 }
 
 /**
- * Actually just get or allocate a free page fragment and arrange fragments
- * array w/o real data splitting.
+ * Get room for @len bytes of data starting from offset @off
+ * in fragment @i.
+ *
+ * The room may be found in the preceding fragment if @off is zero.
+ * Otherwise, a new fragment is allocated and fragments around the
+ * fragment @i are rearranged so that data is not actually split
+ * and copied.
  *
- * @return pointer to the new data room or NULL on failure.
+ * Note: @off is always within the borders of fragment @i. It can
+ * point at the start of a fragment, but it can never point at the
+ * location right after the end of a fragment. In other words, @off
+ * can be zero, but it can not be equal to the size of fragment @i.
+ *
+ * @return pointer to the room for new data, or NULL on failure.
  */
 static void *
 __split_pgfrag_add(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
 		   int len)
 {
-	int split, dst_i = i + !!off;
-	skb_frag_t *frag_dst, *frag = &skb_shinfo(skb)->frags[i];
+	int tail_len;
 	struct sk_buff *skb_dst;
+	skb_frag_t *frag_dst, *frag = &skb_shinfo(skb)->frags[i];
 
+	SS_DBG("%s: skb [%p] i [%d] off [%d] len [%d] fragsize [%d]\n",
+		__func__, skb, i, off, len, skb_frag_size(frag));
 	BUG_ON((off < 0) || (off >= skb_frag_size(frag)));
-	split = off && off < skb_frag_size(frag);
 
 	/*
-	 * Try to append data to a page fragment. If 'off' is zero,
-	 * then try to append data to a preceding page fragment
+	 * Try to append data to the page fragment. If @off is zero,
+	 * then try to append data to the preceding page fragment
 	 * if there's any. Go for other solutions if there's no room.
 	 */
-	if (!split && (off || i)) {
+	if (off || i) {
```

Comment:
The condition is wrong: we're entering to the branch ever if we must insert data into middle of the fragment. Meantime, if shall be ```if (!off && i)```, i.e. try to append the data to previous fragment (i != 0) if we need to add data just before the fragment (!off).

Since now we don't handle case adding data just after i'th fragment, the internals of the branch must also be amended - ```frag_dst``` is always ```frag```, so we actually just should check current ```frag``` and can do this in the condition above.